### PR TITLE
✨ Ranged Unit AI

### PR DESCRIPTION
**Pull request by 429277**
Added a prototype for the ranged (Javelin) unit AI prefab, alongside with corresponding scripts.

### CODE REVIEW

File: Assets/Scripts/Units/Javelin Unit/IJavelinUnitState.cs [hunk 1]
```diff
@@ -0,0 +1,19 @@
+using System.Collections;
+using System.Collections.Generic;
+using UnityEngine;
+
+namespace RogueApeStudio.Crusader.Units.JavelinUnit
+{
+    public enum JavelinUnitStateId
```
Comment (NathanThus) [inline Assets/Scripts/Units/Javelin Unit/IJavelinUnitState.cs:7]: Isn't stateID a bit redundant, as it's an enum?

File: Assets/Scripts/Units/Javelin Unit/Javelin.cs [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+using UnityEngine;
+
+namespace RogueApeStudio
+{
+    public class Javelin : MonoBehaviour
+    {
+        [SerializeField] private Rigidbody _rb;
+        [SerializeField] private Collision _target;
+        [SerializeField] private float _despawnDuration = 1f;
+        [SerializeField] private float _playerLayer = 9;
```
Comment (NathanThus) [inline Assets/Scripts/Units/Javelin Unit/Javelin.cs]: This hardcoded value could change in the future, might want to search for a method that grabs the value.

File: Assets/Scripts/Units/Javelin Unit/JavelinUnit.cs [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+            AddJavelinUnitState(new JavelinUnitChaseState());
+            AddJavelinUnitState(new JavelinUnitAttackState());
+            _currentState = GetJavelinUnitState(JavelinUnitStateId.Chase);
+            LocalUnitMovement.SetStopDistance(_stopDistance);
+            LocalUnitMovement.SetSpeed(_movementSpeed);
+        }
+
+        void Update()
+        {
+            _currentState.UpdateState(this);
```
Comment (NathanThus) [inline Assets/Scripts/Units/Javelin Unit/JavelinUnit.cs:53]: Does this need to run _each_ frame?

File: Assets/Scripts/Units/Javelin Unit/JavelinUnit.cs [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+            return _states[index];
+        }
+
+        public void ChangeState(JavelinUnitStateId stateId)
+        {
+            _currentState = GetJavelinUnitState(stateId);
+            _currentState.EnterState(this);
+        }
+
+        #region Commen Methods
```
Comment (NathanThus) [inline Assets/Scripts/Units/Javelin Unit/JavelinUnit.cs]: Did you mean public?

File: Assets/Scripts/Units/Javelin Unit/JavelinUnitChaseState.cs [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+        public JavelinUnitStateId GetId()
+        {
+            return JavelinUnitStateId.Chase;
+        }
+
+        public void UpdateState(JavelinUnit javelinUnit)
+        {
+            javelinUnit.LocalUnitMovement.MoveToPlayer();
+
+            if (LosTimeHasPassed(javelinUnit))
```
Comment (NathanThus) [inline Assets/Scripts/Units/Javelin Unit/JavelinUnitChaseState.cs]: LOS is not a clear abbrieviation.

**Review by NathanThus — CHANGES_REQUESTED** [review]

File: Assets/Scripts/Units/Javelin Unit/JavelinUnitChaseState.cs (hunk 5 above)
Comment (429277) [reply Assets/Scripts/Units/Javelin Unit/JavelinUnitChaseState.cs]: fixed

File: Assets/Scripts/Units/Javelin Unit/JavelinUnit.cs (hunk 4 above)
Comment (429277) [reply Assets/Scripts/Units/Javelin Unit/JavelinUnit.cs]: fixed

File: Assets/Scripts/Units/Javelin Unit/JavelinUnit.cs (hunk 3 above)
Comment (429277) [reply Assets/Scripts/Units/Javelin Unit/JavelinUnit.cs:53]: yes

File: Assets/Scripts/Units/Javelin Unit/Javelin.cs (hunk 2 above)
Comment (429277) [reply Assets/Scripts/Units/Javelin Unit/Javelin.cs]: fixed

File: Assets/Scripts/Units/Javelin Unit/JavelinUnit.cs (hunk 3 above)
Comment (NathanThus) [reply Assets/Scripts/Units/Javelin Unit/JavelinUnit.cs:53]: I'll make an issue on it, but that's for post-dragons den.

File: Assets/Scripts/Units/Javelin Unit/RogueApeStudios.Units.JavelinUnit.asmdef
Comment (NathanThus) [inline Assets/Scripts/Units/Javelin Unit/RogueApeStudios.Units.JavelinUnit.asmdef:1]: `RogueApeStudio.Crusader.Units.JavelinUnit`

**Review by NathanThus — CHANGES_REQUESTED** [review]

**Review by NathanThus — APPROVED** [review]

**Pull request merged**